[docs] Entra ID integration

flyinghermit opened the pull request:
Adds Entra ID integration docs under Identity Governance:
- Guided OIDC IdP based set up with Web UI and tctl.
- Guided system credential set up with tctl
- Manual OIDC set up with Azure portal clickops and tctl. 

The content is abstracted and augmented from the existing Entra ID docs https://goteleport.com/docs/identity-security/integrations/entra-id/. The primary goal was to put canonical Entra ID docs under Identity Governance with generic usage coverage, which the existing Entra ID doc under Identity Security does not address. 

This PR does not include manual OIDC set up with powershell script. Instead, it adds Azure portal based OIDC configuration steps. Powershell script can be added at a later time.

flyinghermit committed:
guided installation for oidc idp based integration

flyinghermit committed:
add guided step headings

flyinghermit committed:
add manual oidc configuration steps

flyinghermit committed:
copy update

flyinghermit committed:
add faq and next steps

flyinghermit committed:
add guided system credential setup

flyinghermit committed:
Merge branch 'master' into sshah/entra-integration

flyinghermit committed:
add .mdx to links

r0mant commented on docs/pages/identity-governance/entra-id.mdx:
```diff
@@ -0,0 +1,360 @@
+also be enabled if you choose to enable Identity Security integration.
+- Your user must have privileged administrator permissions in the Azure account.
+- For the OIDC IdP setup, the Teleport cluster must be publicly accessible from the internet.
+- For air gapped clusters, `tctl` must be v16.4.7 or later.
```
Nit: I'd consider removing this line. This guide will only apply to v18+ so a user would likely have a newer version anyway and v16 will be obsolete soon.
```suggestion
```

r0mant reviewed (approved):
Thank you for putting this together. Looks like a solid guide, don't have much feedback :shipit:

flyinghermit committed:
Update docs/pages/identity-governance/entra-id.mdx

Co-authored-by: Roman Tkachenko <roman@goteleport.com>

tigrato commented on docs/pages/identity-governance/entra-id.mdx:
```diff
@@ -0,0 +1,359 @@
+2. **User sync:** Imports Entra ID users as Teleport users.
+3. **Group sync:** Imports Entra ID groups as Teleport
+   Access Lists. 
+4. **Application sync (Optional):** Imports Entra ID applications. 
```
We do not import Applications into teleport. We only import them to access graph.
As opposed to Okta, we only import groups & users into Teleport

tigrato commented on docs/pages/identity-governance/entra-id.mdx:
```diff
@@ -0,0 +1,359 @@
+3. **Group sync:** Imports Entra ID groups as Teleport
+   Access Lists. 
+4. **Application sync (Optional):** Imports Entra ID applications. 
+5. **Integration with Teleport Identity Security (Optional):** Lets you analyze user access paths
```
Refer that we also import any enterprise app to be displayed there

tigrato commented on docs/pages/identity-governance/entra-id.mdx:
```diff
@@ -0,0 +1,359 @@
+An enterprise application will be configured in the Entra ID. 
+The purpose of this application is twofold - 
+1. Set up a SAML SSO provider for Teleport. 
+2. Set up Microsoft graph API credential for Teleport that allows to fetch resources 
```
This is not accurate if using system credentials and we should enphasize that here to avoid confusion. 

The second case is only true if using OIDC for authentication, otherwise that's not required at all.

tigrato commented on docs/pages/identity-governance/entra-id.mdx:
```diff
@@ -0,0 +1,359 @@
+from Entra ID directory to Teleport.  
+
+Once the integration is configured, Teleport periodically imports users, groups, 
+and applications into Teleport.
```
applications are not imported

tigrato commented on docs/pages/identity-governance/entra-id.mdx:
```diff
@@ -0,0 +1,359 @@
+
+In this setup, Teleport is configured as an OpenID Connect (OIDC) identity provider, 
+establishing a trusted connection with an Entra ID application created during setup. 
+This trust allows Teleport to authenticate using the Entra ID application, accessing 
```
```suggestion
This trust allows Teleport to generate short-lived credentials to authenticate using the Entra ID application, accessing 
```

tigrato commented on docs/pages/identity-governance/entra-id.mdx:
```diff
@@ -0,0 +1,359 @@
+
+Designed for air-gapped Teleport clusters that are not publicly accessible, this setup 
+accommodates environments where Azure cannot validate OIDC tokens issued by Teleport. 
+Instead, Teleport relies on Azure credentials available on the VMs where Teleport 
```
I think we should rephrase this to avoid referring VMs, first because it works with kubernetes pods and second because we can make those credentials available in AWS/GCP machines

tigrato commented on docs/pages/identity-governance/entra-id.mdx:
```diff
@@ -0,0 +1,359 @@
+
+- Teleport Identity Governance enabled for your account. Optionally, Identity Security should
+also be enabled if you choose to enable Identity Security integration.
+- Your user must have privileged administrator permissions in the Azure account.
```
```suggestion
- Your user must have privileged administrator permissions in the Azure EntraID account.
```

This requires Admin for EntraID, not azure

tigrato commented on docs/pages/identity-governance/entra-id.mdx:
```diff
@@ -0,0 +1,359 @@
+You will need to grant Azure Identity with the necessary permissions required for the Entra ID
+integration.
+
+In the Azure Portal, find the identities linked to your Teleport Auth Service VMs,
```
lets rephrase this with the comments I left before about kube pods and other vms

tigrato commented on docs/pages/identity-governance/entra-id.mdx:
```diff
@@ -0,0 +1,359 @@
+The integration is now configured and the Teleport Entra ID service will start 
+importing resources from Entra ID to Teleport.
+
+## Manual configuration
```
This section of the manual is confusing because it's part of the main flow of the documentation, yet it replaces-rather than complements-the other methods described in the "Pages" section. It's also unclear why it starts from step 1/x, which adds to the confusion.

zmb3 replied:
Agreed, I don't think we've nailed the UX here yet. This is very hard to follow.

tigrato commented on docs/pages/identity-governance/entra-id.mdx:
```diff
@@ -0,0 +1,359 @@
+![Entra ID Integration](../../img/igs/entraid/entra-sso-config.png)
+
+
+## Step 3/5. Configure OIDC IdP
```
The policy integration also shows the powershell script to create the app and OIDC. This is important if you want to have some kind of reproducible deployment that automatically creates everything in entra id side.

I think you dropped them here but I would rather include them

flyinghermit replied:
Its dropped temporarily cause I haven't tested those scripts and wanted to reduce number of changes in this PR. I think its fine to follow up with that later as long as the manual steps in the UI is covered.

zmb3 commented on docs/pages/identity-governance/entra-id.mdx:
```diff
@@ -0,0 +1,359 @@
+Open the newly created enterprise application.
+
+Under “Manage” menu select “Single sign-on”. In this configuration UI, you will need to set up Teleport
+as an SAML servicer provider.
```
I'm confused - is this setup using both OIDC and SAML?

zmb3 commented on docs/pages/identity-governance/entra-id.mdx:
```diff
@@ -0,0 +1,359 @@
+- **Federated credential scenario:** Other issuer
+
+Under “Connect your account”, configure the following values:
+- **Issuer:** https://example.com
```
Someone reading these docs is likely to type the literal "example.com" value in here, which would be wrong. We need to better describe the steps we want the user to take.

zmb3 commented on docs/pages/identity-governance/entra-id.mdx:
```diff
@@ -0,0 +1,359 @@
+You can add a new graph permission by clicking on `+ Add a permission` button and then selecting 
+“Microsoft Graph > Application permissions”.
+
+The following permissions needs to be added to the application. 
```
```suggestion
The following permissions need to be added to the application. 
```

zmb3 commented on docs/pages/identity-governance/entra-id.mdx:
````diff
@@ -0,0 +1,359 @@
+    –-manual-setup
+```
+
+`tctl` will then prompt for Entra ID tenant ID and Application ID of the enterprise 
````
Why is tenant ID lowercase but Application ID is capitalized?

flyinghermit replied:
Only because I failed to notice that 😢

zmb3 commented on docs/pages/identity-governance/entra-id.mdx:
````diff
@@ -0,0 +1,359 @@
+- User.Read.All
+```
+
+If you enable Identity Security integration, you will need a broader sets of permission.
````
```suggestion
If you enable the Identity Security integration, you will need a broader set of permissions.
```

Do these permissions _replace_ the ones above, or are they needed in addition to the ones above?

flyinghermit replied:
Replace it. Added comment to hint that.

zmb3 reviewed (changes requested):
Good start but I think this needs a bit of work. I found it difficult to follow with all the different options and flows, with little guidance as to how to choose which option is most appropriate.

r0mant reviewed (commented):
@flyinghermit To make this guide simpler and easier to follow I'm thinking let's remove all Identity Security specific stuff from it, things like importing applications, analyzing paths, etc. Otherwise users who are only looking at it for IG might get confused. What I'd do instead, is to add a link to the respective Identity Security guide in the "Next Steps" section.

flyinghermit committed:
review comments

flyinghermit committed:
simplify guided configuration steps

flyinghermit committed:
add id sec link, update command ref"

flyinghermit committed:
briefly explain flags

flyinghermit commented on docs/pages/identity-governance/entra-id.mdx:
````diff
@@ -0,0 +1,359 @@
+To begin integration, run the `tctl plugins install entraid` command.
+
+```code
+$ tctl plugins install entraid \
````
I wanted to have them explained in a followup PR where I will be adding the entra ID plugins command in the tctl reference page.  Added a brief descriptions for each flag in https://github.com/gravitational/teleport/pull/56930/commits/bcf67a2e15836f0bb2756c42a45528d9b86a1f77

flyinghermit committed:
refactor: split the guide into getting started and manual pages

r0mant commented on docs/pages/identity-governance/entra-id/getting-started.mdx:
```diff
@@ -0,0 +1,223 @@
+The integration is now configured and the Teleport Entra ID service will start 
+importing resources from Entra ID to Teleport.
+
+## FAQ
```
Should we move FAQ into its own .mdx file and link to it in the next steps? Because you could have those questions after following either of the guides. Same for "Access Management" section we'll add next, I think it should be a separate page under Entra section.

r0mant commented on docs/pages/identity-governance/entra-id/manual-installation.mdx:
```diff
@@ -0,0 +1,104 @@
+`tctl` will then prompt for Entra ID tenant ID and application ID of the enterprise 
+application created in step 1.
+
+After you enter these values, Entra ID plugin will be installed with the OIDC IdP based authentication method.
```
I would put the same "next steps" here as in the guided flow.

r0mant reviewed (approved):
This is much easier to follow, just have a couple of comments.

As we already discussed, once this merges, as a next step let's extend these guides with:

- Showing how to create Azure resources with terraform. Similar to what you did here, I think a separate guide would be appropriate? So we don't complicate the "manual" one.
- Add a page that explains RBAC, for example after I have setup Entra sync, how do I give access to some Teleport resources to my Entra groups?

zmb3 commented on docs/pages/identity-governance/entra-id/entra-id.mdx:
```diff
@@ -0,0 +1,72 @@
+To configure SSO, Teleport uses an Entra ID enterprise application where 
+Teleport must be set up as an SAML service provider. 
+
+To import users and groups from the Entra ID, Teleport must be configured with 
```
```suggestion
To import users and groups from Entra ID, Teleport must be configured with 
```

zmb3 commented on docs/pages/identity-governance/entra-id/entra-id.mdx:
```diff
@@ -0,0 +1,72 @@
+
+1. **Single Sign-On (SSO):** Configures Teleport authentication with Entra ID as an
+   identity provider.
+2. **User sync:** Periodic import of the Entra ID users as Teleport users.
```
```suggestion
2. **User sync:** Periodic import of Entra ID users as Teleport users.
```

zmb3 commented on docs/pages/identity-governance/entra-id/entra-id.mdx:
```diff
@@ -0,0 +1,72 @@
+1. **Single Sign-On (SSO):** Configures Teleport authentication with Entra ID as an
+   identity provider.
+2. **User sync:** Periodic import of the Entra ID users as Teleport users.
+3. **Group sync:** Periodic import of the Entra ID groups as Teleport
```
```suggestion
3. **Group sync:** Periodic import of Entra ID groups as Teleport
```

zmb3 commented on docs/pages/identity-governance/entra-id/entra-id.mdx:
```diff
@@ -0,0 +1,72 @@
+#### Teleport as an OIDC Provider for Entra ID 
+
+In this setup, Teleport is configured as an OpenID Connect (OIDC) identity provider 
+for the Entra ID enterprise application, which is created during the setup. 
```
```suggestion
for the Entra ID enterprise application.
```

zmb3 commented on docs/pages/identity-governance/entra-id/entra-id.mdx:
```diff
@@ -0,0 +1,72 @@
+In this setup, Teleport is configured as an OpenID Connect (OIDC) identity provider 
+for the Entra ID enterprise application, which is created during the setup. 
+Teleport OIDC IdP then generates a short-lived credential for the Microsoft Graph API
+client configured for the Entra ID. Authorization is limited to the API permission 
```
```suggestion
client configured for Entra ID. Authorization is limited to the API permission 
```

zmb3 commented on docs/pages/identity-governance/entra-id/entra-id.mdx:
```diff
@@ -0,0 +1,72 @@
+configured in the Entra ID enterprise application. 
+
+Direct bidirectional connectivity between Teleport and Entra ID is necessary for 
+the Entra ID to validate the OIDC tokens issued by Teleport.
```
```suggestion
Entra ID to validate the OIDC tokens issued by Teleport.
```

zmb3 commented on docs/pages/identity-governance/entra-id/entra-id.mdx:
```diff
@@ -0,0 +1,72 @@
+Designed for air-gapped Teleport clusters that are not publicly accessible, this setup 
+accommodates environments where Entra ID cannot validate OIDC tokens issued by Teleport. 
+
+Instead, Teleport relies on API credentials available where Teleport 
```
```suggestion
Instead, Teleport relies on API credentials available where the Teleport 
```

zmb3 commented on docs/pages/identity-governance/entra-id/entra-id.mdx:
```diff
@@ -0,0 +1,72 @@
+script, which configures your Entra ID tenant with the properties that are required
+for the Teleport Entra ID integration. 
+
+But if you want to have more control over the Entra ID configuration, a manual
```
```suggestion
If you want to have more control over the Entra ID configuration, a manual
```

zmb3 commented on docs/pages/identity-governance/entra-id/getting-started.mdx:
```diff
@@ -0,0 +1,223 @@
+
+## Prerequisites
+
+- Teleport Identity Governance enabled for your Teleport cluster. Optionally, Identity Security should
+also be enabled if you choose to enable Identity Security integration.
```
If we removed the identity security info from this guide, we can probably remove the prereq.

zmb3 commented on docs/pages/identity-governance/entra-id/manual-installation.mdx:
```diff
@@ -0,0 +1,104 @@
+ - identity-governance
+---
+
+This guide shows manual Entra ID configuration steps to set up Teleport Entra ID integration.
```
It might be helpful to add one more sentence to the intro that cross-links to the other guide.

Something like:

> For instructions on the guided installation, see [other guide] instead.

zmb3 reviewed (approved):
Looks good, only minor suggestions.

flyinghermit committed:
review suggestions:
- add a dedicated faq,
- add next steps to the manual guide
- remove residual access graph reference
- grammar fixes

flyinghermit commented:
Thanks for review. I've resolved all the suggestions with this commit https://github.com/gravitational/teleport/pull/56930/commits/6211d813b69c4b963bfcc971299a181e1c9b7b14.

Nested access list, rbac, terraform guide etc will be followed up separately.

flyinghermit committed:
Merge branch 'master' into sshah/entra-integration
